### DISCUSSION TITLE
Modules in python Cadwork plugin

### DISCUSSION THREAD

**Discussion by usai-sylvain** [cwapi3d/cwapi3dpython#118] [Q&A] [answered]
I am trying to organize my python plugin into modules. Ideally having a minimal "_PLUGIN_NAME_.py" that just call logic defined in subfolders. 

![image](https://github.com/cwapi3d/cwapi3dpython/assets/3482224/23a996f4-51cd-43bc-b75b-4ddafe5d7480)

```python
import time

print("sandbox start")

from Sandbox import TestObject

TestObject.Display()

print("sandbox end")

time.sleep(1)
print("bye")
time.sleep(1)
```

![image](https://github.com/cwapi3d/cwapi3dpython/assets/3482224/63efce3b-6865-4f4b-aecd-6d1ae5928537)

Running SUS_Sandbox.py prints results as expected. But running it in Cadwork leads to import issues : 

![image](https://github.com/cwapi3d/cwapi3dpython/assets/3482224/fbb2cc55-a96c-46a9-b1c0-c3671ae55116)

Anyclue how to solve my imports ? 
Thanks !

edit : added code : [SUS_Sandbox.zip](https://github.com/cwapi3d/cwapi3dpython/files/14037955/SUS_Sandbox.zip)

**Comment by bastiannispel** [comment]
Die you import "Sandbox" (which is a submodule) in the __init__ of the main package ?

**Reply by usai-sylvain** [reply]
Thanks Basti, 
Yes, I tried that but I still get the same error...

**Comment by Brunner246** [comment]
Hi @usai-sylvain 
You must make the path to the dependecies known.

See => https://docs.cadwork.com/projects/cwapi3dpython/en/latest/modules/

```python
import sys

import cadwork
import element_controller
import  utility_controller as uc
plugin_path = uc.get_plugin_path()  

path_parts = [plugin_path , '.venv', 'Lib', 'site-packages']
path = '/'.join(path_parts)
sys.path.append(path)

import sandbox...
```

**Reply by usai-sylvain** [reply]
both 

```python
import time
import sys
import os

import  utility_controller as uc
# get userprofil path
USERPROFIL = uc.get_3d_userprofil_path()   

# appending a path
sys.path.append(USERPROFIL + '\\API.x64\\SUS_Sandbox\\SandboxObjects')


for p in sys.path:
    print(p)

print ("try to import now")
from SandboxObjects import TestObject
```


![image](https://github.com/cwapi3d/cwapi3dpython/assets/3482224/81c50fa0-ad37-4272-9d83-644976a6a8da)

and 

```python
import time
import sys
import os

import element_controller
import  utility_controller as uc
plugin_path = uc.get_plugin_path()  

path_parts = [plugin_path , 'SandboxObjects']
path = '/'.join(path_parts)
sys.path.append(path)

for p in sys.path:
    print(p)

print ("try to import now")
from SandboxObjects import TestObject
```

![image](https://github.com/cwapi3d/cwapi3dpython/assets/3482224/2f163800-050a-44df-bea0-ba01df8a274a)

give the same error. (I renamed the folder "SandboxObjects" in the meantime to avoid naming conflict with pyproject...)

**Comment by Brunner246** [comment, marked as answer]
@usai-sylvain 

this works https://github.com/Brunner246/cwapi3d_examples/blob/master/github_questions/SUS_Sandbox.py

**Reply by usai-sylvain** [reply]
Thanks Michael !! it does run. Not super happy about having to declare all method as import but that's definitively a good start. 
Café's on me next time you are in zurich :) (@bastiannispel we already share a coffee machine, but I can press the button for you next time :p )

**Reply by Brunner246** [reply]
you could also do this 
```python
import sys

print("sandbox start")

import utility_controller as uc

plugin_path = uc.get_plugin_path()

sys.path.append(plugin_path)

import Sandbox.TestObject

Sandbox.TestObject.Display()

print("sandbox end")
````

**Reply by usai-sylvain** [reply]
making TestObject a class allow to import the entire class and access all methods within. I can def work with that ! problem solved :)
